File: things/temperature_sensor.py
"""
Temperature sensor class.
"""
import random

from thing import Thing


class TemperatureSensor(Thing):
    """
    Represents a temperature sensor.
    """
    _precision: int = 2
    _minimum: float = -20.0
    _maximum: float = 40.0

    _value: float = float("nan")

    def value(self) -> float:
        """Get the value of the temperature sensor

        Returns:
            float: The value of the temperature sensor
        """
        return self._value

    async def _update(self) -> None:
        self._value = await self._read_value()

    async def _read_value(self) -> float:
        """
        Simulate reading the value of the temperature sensor.
        """
        return round(random.uniform(self._minimum, self._maximum), self._precision)

    def get_info(self) -> str:
        """
        Get information about the temperature sensor.
        """
        return "Temperature sensor 9000"

File: things/temperatureSensor.ts
import { sleep } from "./helpers";
import { Sensor } from "./thing";

export class TemperatureSensor extends Sensor {
    /**
     * The temperature as read by the sensor in degrees Celsius
     */
    temperature: number;

    constructor() {
        super();
        this.temperature = NaN;
    }

    async tick(): Promise<boolean> {
        const value = Math.floor(Math.random() * 100);
        this.temperature = value;

        await sleep(1 * 1000); // Busy work

        return true;
    }

    /**
     * Get a human-readable interpretation of the temperature value from the raw sensor data
     * @example
     * // returns ['Temperature: 1 °C', 'Temperature: 2 °C']
     * interpretResults([1, 2])
     * @param results The results of the temperature sensor readings, as from the value getter
     * @returns A human-readable interpretation of the results
     */
    public interpretResults(results: number[]): string[] {
        const intepretations: string[] = results.map(
            (el) => `Temperature: ${el} °C`
        );
        return intepretations;
    }

    public get value(): number {
        return this.temperature;
    }
}
